poseidon/src/lib.rs

```diff
@@ -0,0 +1,193 @@
+mod constants;
+
+use constants::{
+    RATE,
+    ROUNDS_FULL,
+    ROUNDS_PARTIAL,
+    ROUND_KEYS_COMPRESSED,
+    WIDTH, // total width of the state
+};
+
+use ark_ff::{BigInt, Field};
+use felt::Felt252;
+use std::marker::PhantomData;
+
+pub trait FieldHasher<F: Field> {
+    fn hash(input: &[F]) -> F;
+
+    fn pair(left: F, right: F) -> F;
+}
+
+pub struct Poseidon3<F: Field> {
+    _ph: PhantomData<F>,
+}
+
+#[inline]
+fn s_box<F: Field>(x: F) -> F {
+    x.square() * x
+}
+
+impl Poseidon3<Felt252> {
+    fn mix(st: &mut [Felt252; WIDTH]) {
+        // linear layer: mix state
+        // M = (
+        //  ( 3, 1, 1),
+        //  ( 1,-1, 1),
+        //  ( 1, 1, -2)
+        // )
+        // see:
+        // https://github.com/starkware-industries/poseidon/blob/main/poseidon3.txt
+        let t0 = st[0] + st[1];
+        let t1 = t0 + st[2];
+        st[0] = t1 + st[0].double();
+        st[1] = t1 - st[1].double();
+        st[2] = t0 - st[2].double();
```

Comment:
Compare to https://github.com/lambdaclass/lambdaworks/blob/main/crypto/src/hash/poseidon/starknet/parameters.rs#L25
You reduced one addition operation!

It might be good to submit a PR to the lambdaworks implementation. (Is there a conflict of interest?)